File: work/dp_f.py
# -*- coding: utf-8 -*-
s = input()
ns = len(s)
t = input()
nt = len(t)

dp = [[0 for _ in range(ns+1)] for _ in range(nt+1)]

for ti in range(nt):
    for si in range(ns):
        if t[ti]==s[si]:
            add = 1
        else:
            add = 0

        dp[ti+1][si+1] = max(
            dp[ti][si]+add,
            dp[ti+1][si],
            dp[ti][si+1]
        )

# print(dp[nt][ns])

res = []
ti = nt
si = ns
while ti>0 and si>0:
    if t[ti-1]==s[si-1]:
        res.append(s[si-1])
        si -= 1
        ti -= 1
    elif dp[ti][si-1]>dp[ti-1][si]:
        si -= 1
    else:
        ti -= 1

res.reverse()
print("".join(res))

File: work/dp_f.cpp
#include<iostream>
#include<string>
#include<vector>
using namespace std;

int main(){
  string s,t;
  cin>>s>>t;
  int ns = s.length();
  int nt = t.length();

  vector<vector<int>> dp =
    vector<vector<int>>(nt+1, vector<int>(ns+1,0));

  for(int ti=0;ti<nt;ti++){
    for(int si=0;si<ns;si++){
      int add = 0;
      if(t[ti]==s[si]){
        add = 1;
      }

      dp[ti+1][si+1] = max(
        dp[ti][si]+add,
        max(
          dp[ti+1][si],
          dp[ti][si+1]
        )
      );
    }
  }

  string res = "";
  int ti = nt;
  int si = ns;
  while(ti>0 && si>0){
    if(t[ti-1]==s[si-1]){
      res = s[si-1]+res;
      si -= 1;
      ti -= 1;
    }else if(dp[ti][si-1]>dp[ti-1][si]){
      si -= 1;
    }else{
      ti -= 1;
    }
  }

  cout<<res<<endl;
  return 0;
}
